### PR TITLE
feat: support multiple ionic projects

### PR DESCRIPTION
**Pull request by stupidawesome**
This PR adds experimental support for multiple Ionic apps within a shared workspace by allowing separate cordova integrations to be configured for each project. This is achieved by augmenting the `ionic.config.json` object with new properties in a way that is backwards compatible with existing single app projects. This is currently limited to `angular` project types with `cordova` integrations only.

Related issue: #3281 

Tests should be passing but I need to write some more that cover multi-app scenarios.

### CODE REVIEW

File: packages/@ionic/cli-utils/src/lib/project/index.ts [hunk 1]
```diff
@@ -31,39 +50,40 @@ export interface ProjectDeps {
...
-    if (projectFile && projectFile.type && PROJECT_TYPES.includes(projectFile.type)) {
-      debug(`Project type from config: ${chalk.bold(prettyProjectName(projectFile.type))} ${projectFile.type ? chalk.bold(`(${projectFile.type})`) : ''}`);
-      return projectFile.type;
+    if (type && PROJECT_TYPES.includes(type)) {
+      debug(`Project type from config: ${chalk.bold(prettyProjectName(type))} ${type ? chalk.bold(`(${type})`) : ''}`);
+      return type;
     }
 
     for (const projectType of PROJECT_TYPES) {
-      const p = await Project.createFromProjectType(projectDir, PROJECT_FILE, deps, projectType);
```
[the remark this answers is not in the document]
Comment (stupidawesome) [reply packages/@ionic/cli-utils/src/lib/project/index.ts]: Try this:

ionic.config.json
```json
{
  "defaultProject": "sidemenu",
  "projects": {
    "sidemenu": {
      "type": "angular"
    },
    "tabs": {
      "type": "angular"
    }
  }
}
```
angular.json
```json
{
  "defaultProject": "sidemenu",
  "projects": {
    "sidemenu": {
      "root": "packages/sidemenu"
    },
    "tabs": {
      "root": "packages/tabs"
    }
  }
}
```

File: packages/ionic/src/commands/link.ts [hunk 2]
```diff
@@ -197,11 +197,13 @@ ${chalk.cyan('[2]')}: ${chalk.bold('https://ionicframework.com/support/request')
           type: 'input',
           name: 'name',
           message: 'Please enter a name for your new app:',
-          validate: v => validators.required(v),
+          validate: (v: string) => validators.required(v),
         });
       }
 
-      proId = await this.createApp({ name }, runinfo);
+      if (name) {
```
[the remark this answers is not in the document]
Comment (stupidawesome) [reply packages/ionic/src/commands/link.ts]: I think I was getting some TypeScript validation error, but I'm not seeing it now so I'll revert it.

File: packages/ionic/src/commands/start.ts [hunk 3]
```diff
@@ -434,7 +434,7 @@ ${chalk.cyan('[1]')}: ${chalk.bold('https://ionicframework.com/docs/cli/starters
     // start is weird, once the project directory is created, it becomes a
     // "project" command and so we replace the `Project` instance that was
     // autogenerated when the CLI booted up. This has worked thus far?
-    this.env.project = await getProject(projectDir, this.env);
+    this.env.project = await getProject(projectDir, String(options['name']), this.env);
```
[the remark this answers is not in the document]
Comment (stupidawesome) [reply packages/ionic/src/commands/start.ts]: Yeah I wasn't sure what to put here, is this after the project is created? Will the name entered in `ionic start` also be used as the project name for the app that gets generated? Or will they always default to `app`?

File: packages/@ionic/cli-utils/src/definitions.ts [hunk 4]
```diff
@@ -98,6 +99,15 @@ export interface ProjectFile {
...
     cert?: string;
   };
+
+  type: ProjectType;
+}
+
+export interface MultiProjectConfig {
+  defaultProject: string;
+  projects: {
+    [key: string]: ProjectConfig;
```
[the remark this answers is not in the document]
Comment (stupidawesome) [reply packages/@ionic/cli-utils/src/definitions.ts]: I added the type guards `isProjectConfig` and `isMultiProjectConfig()` to check if the `projects` key exists and toggle behaviour accordingly.

File: packages/@ionic/cli-utils/src/lib/project/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dwieeb) [reply packages/@ionic/cli-utils/src/lib/project/index.ts]: Sorry, I was just trying to convey that `determineType()` doesn't work for multi-app projects. I'm thinking the function needs to be refactored to either work or throw a fatal error.

Remove the `type` attributes from `ionic.config.json` to see what I mean.

File: packages/@ionic/cli-utils/src/lib/build.ts [hunk 5]
```diff
@@ -26,6 +39,11 @@ export const COMMON_BUILD_COMMAND_OPTIONS: ReadonlyArray<CommandMetadataOption>
     summary: `Target platform on chosen engine (e.g. ${['ios', 'android'].map(e => chalk.green(e)).join(', ')})`,
     groups: [OptionGroup.Advanced],
   },
+  {
+    name: 'project',
+    summary: 'The name of the project',
```
[the remark this answers is not in the document]
Comment (dwieeb) [reply packages/@ionic/cli-utils/src/lib/build.ts]: I need to do some refactoring for these commands. Maybe have "runners" for each command/project type combination. Then it'd be easy to add the option to the project type(s) that support global options like this.

File: packages/@ionic/cli-utils/src/definitions.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (dwieeb) [reply packages/@ionic/cli-utils/src/definitions.ts]: But still if I access `projectConfig.projects.asndgksjehbnrgfvkljahsefdkjghasekjhfgakjhsdg` then TypeScript will say it's a `ProjectConfig`, when it possibly is not there at all.

File: packages/@ionic/cli-utils/src/lib/project/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (stupidawesome) [reply packages/@ionic/cli-utils/src/lib/project/index.ts]: Should this perhaps throw an error instead?

https://github.com/ionic-team/ionic-cli/blob/master/packages/%40ionic/cli-utils/src/lib/project/index.ts#L78

[the remark this answers is not in the document]
Comment (stupidawesome) [reply packages/@ionic/cli-utils/src/lib/project/index.ts]: I added more exceptions so it should give more meaningful error messages now.

[the remark this answers is not in the document]
Comment (dwieeb) [reply packages/@ionic/cli-utils/src/lib/project/index.ts]: Traditionally `determineType()` wasn't meant to throw fatal errors. If a project type cannot be determined, it's actually fine, it just means we can't use project commands. The CLI switches to "global mode" if project type cannot be determined.

But if the behavior is changed, it's fine. We don't have integration tests for behavior, so it's mostly going to be me testing things and making tweaks. 😆

**Comment by kevin-lot** [comment]
I am very happy to see this PR merged.
Are you planning to add the same spirit to capacitor client ?

(and how can we use the client with this new feature ?)

**Comment by timbru31** [comment]
FYI this does not work with the Snapshot Build of Ionic Pro (yet). I've opened a support ticket to have this issue addressed.  
Change is awesome nevertheless! 👍

**Comment by dwieeb** [comment]
@kevin-lot Yes, this will work for Capacitor.

@timbru31 Thanks for the heads up. I've been working with @nphyatt to get this feature working in Ionic Pro by using the Ionic CLI for builds. Note that this will only be for Ionic 4 beta, so some rough spots are to be expected. 😊

**Comment by edcarroll** [comment]
I've been working with the new multi project setup and it's been fantastic, though have run into an issue: when running any Cordova commands on Windows, the use of an absolute path in `--cordova-base-path` stops the Angular CLI from actually outputting the built code in the correct place, which then makes `cordova prepare` fail as the `www` folder is missing.

Using a relative path fixes this (I believe it's when the absolute path is normalized, which converts `C:\` to `/C/`) - though I've not yet worked out exactly where this absolute path is coming from. Hoping to open a PR when I find it, but if anyone spots this and knows where to go that would be very helpful!

**Comment by dwieeb** [comment]
@edcarroll Interesting. I wonder if the Angular Devkit has issues with using an absolute path within a workspace. There are levels of abstraction with paths and it may get confused on Windows.

If you're willing to look into it (I would love a PR ❤️), check out these points in the code base:

Ionic CLI provides `--cordova-base-path` option here:

https://github.com/ionic-team/ionic-cli/blob/2862762450b1aeca4b3a0b9bd6165025de9bd3dd/packages/%40ionic/cli-utils/src/lib/project/angular/build.ts#L109

ng-toolkit (using Angular Devkit) picks up the option and normalizes the path here:

https://github.com/ionic-team/ionic-cli/blob/49d9493b2d8ebe02f9e1e19f3be66b4d7d4d2bf6/packages/%40ionic/ng-toolkit/src/cordova-build/index.ts#L37

(also look for the same option in the serve code for Cordova livereload as well)

Maybe the solution is to use a relative path, because it's likely what the Angular Devkit expects within a workspace.

**Pull request merged**